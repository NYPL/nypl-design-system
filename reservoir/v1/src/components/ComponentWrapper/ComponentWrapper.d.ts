import React from "react";
import { HelperErrorTextType } from "../HelperErrorText/HelperErrorText";
export interface ComponentWrapperProps {
    /** The UI elements that will be wrapped by this component */
    children: React.ReactNode;
    /** A class name for the `div` parent element. */
    className?: string;
    /** Optional string to set the text for the component's description */
    descriptionText?: string | JSX.Element;
    /** Optional string to set the text for a `Heading` component */
    headingText?: string;
    /** Optional string to set the text for a `HelperErrorText` component */
    helperText?: HelperErrorTextType;
    /** Styles that target the helper text. */
    helperTextStyles?: {
        [key: string]: any;
    };
    /** ID that other components can cross reference for accessibility purposes */
    id?: string;
    /** Optional string to populate the `HelperErrorText` for the error state
     * when `isInvalid` is true. */
    invalidText?: HelperErrorTextType;
    /** Sets invalid text in the error state. */
    isInvalid?: boolean;
    /** Offers the ability to hide the helper/invalid text. */
    showHelperInvalidText?: boolean;
}
export declare const ComponentWrapper: import("@chakra-ui/react").ChakraComponent<React.ForwardRefExoticComponent<ComponentWrapperProps & {
    children?: React.ReactNode;
} & React.RefAttributes<HTMLDivElement>>, {}>;
export default ComponentWrapper;
