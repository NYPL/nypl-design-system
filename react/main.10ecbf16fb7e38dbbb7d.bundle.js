(window.webpackJsonp=window.webpackJsonp||[]).push([[0],{1292:function(n,o,c){"use strict";c.r(o);c(1293),c(1323),c(1335),c(1442),c(1485)},1448:function(n,o){},1449:function(n,o){},1450:function(n,o){},1451:function(n,o){},501:function(n,o,c){c(502),c(645),n.exports=c(1292)},564:function(n,o){}},[[501,1,2]]]);