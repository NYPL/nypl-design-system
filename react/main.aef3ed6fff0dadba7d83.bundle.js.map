{"version":3,"file":"main.aef3ed6fff0dadba7d83.bundle.js","sources":["webpack:///main.aef3ed6fff0dadba7d83.bundle.js"],"mappings":"AAAA","sourceRoot":""}