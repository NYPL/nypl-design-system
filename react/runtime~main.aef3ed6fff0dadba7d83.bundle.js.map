{"version":3,"file":"runtime~main.aef3ed6fff0dadba7d83.bundle.js","sources":["webpack:///runtime~main.aef3ed6fff0dadba7d83.bundle.js"],"mappings":"AAAA","sourceRoot":""}