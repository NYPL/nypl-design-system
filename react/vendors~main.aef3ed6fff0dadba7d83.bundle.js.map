{"version":3,"file":"vendors~main.aef3ed6fff0dadba7d83.bundle.js","sources":["webpack:///vendors~main.aef3ed6fff0dadba7d83.bundle.js"],"mappings":";AAAA","sourceRoot":""}