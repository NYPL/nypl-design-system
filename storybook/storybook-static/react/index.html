<!doctype html><html lang="en"><head><meta charset="utf-8"><title>Storybook</title><meta name="viewport" content="width=device-width,initial-scale=1"><style>html, body {
    overflow: hidden;
    height: 100%;
    width: 100%;
    margin: 0;
    padding: 0;
  }</style><script>/* globals window */
  /* eslint-disable no-underscore-dangle */
  try {
    if (window.top !== window) {
      window.__REACT_DEVTOOLS_GLOBAL_HOOK__ = window.top.__REACT_DEVTOOLS_GLOBAL_HOOK__;
    }
  } catch (e) {
    // eslint-disable-next-line no-console
    console.warn('unable to connect to top frame for connecting dev tools');
  }</script><style>#root[hidden],
      #docs-root[hidden] {
        display: none !important;
      }</style></head><body><div id="root"></div><div id="docs-root"></div><script>window['DOCS_MODE'] = false;</script><script src="runtime~main.00bd9d033d7a0b972fc9.bundle.js"></script><script src="vendors~main.d2b6aafef1a259d165c0.bundle.js"></script><script src="main.3bf71672a6dd35499004.bundle.js"></script></body></html>