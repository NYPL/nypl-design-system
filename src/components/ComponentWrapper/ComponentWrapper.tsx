import { Box, chakra, useMultiStyleConfig } from "@chakra-ui/react";
import React, { forwardRef } from "react";

import Heading from "../Heading/Heading";
import HelperErrorText, {
  HelperErrorTextType,
} from "../HelperErrorText/HelperErrorText";
import Text from "../Text/Text";
import Tooltip from "../Tooltip/Tooltip";

export interface ComponentWrapperProps {
  /** The UI elements that will be wrapped by this component */
  children: React.ReactNode;
  /** A class name for the `div` parent element. */
  className?: string;
  /** Optional string to set the text for the component's description */
  descriptionText?: string | JSX.Element;
  /** Optional string to set the text for a `Heading` component */
  headingText?: string;
  /** Optional string to set the text for a `HelperErrorText` component */
  helperText?: HelperErrorTextType;
  /** Styles that target the helper text. */
  helperTextStyles?: { [key: string]: any };
  /** ID that other components can cross reference for accessibility purposes */
  id?: string;
  /** Optional string to populate the `HelperErrorText` for the error state
   * when `isInvalid` is true. */
  invalidText?: HelperErrorTextType;
  /** Sets invalid text in the error state. */
  isInvalid?: boolean;
  /** Offers the ability to hide the helper/invalid text. */
  showHelperInvalidText?: boolean;
  tooltipLabelText?: any;
}

export const ComponentWrapper = chakra(
  forwardRef<HTMLDivElement, React.PropsWithChildren<ComponentWrapperProps>>(
    (props, ref?) => {
      const {
        children,
        className,
        descriptionText,
        headingText,
        helperText,
        helperTextStyles = {},
        id,
        invalidText,
        isInvalid = false,
        showHelperInvalidText = true,
        tooltipLabelText,
        ...rest
      } = props;
      const hasChildren = !!children;
      const styles = useMultiStyleConfig("ComponentWrapper", { hasChildren });
      const footnote = isInvalid ? invalidText : helperText;

      // Note: Typescript warns when there are no children passed and
      // doesn't compile. This is meant to log in non-Typescript apps.
      if (!hasChildren) {
        console.warn(
          "NYPL Reservoir ComponentWrapper: No children were passed."
        );
      }

      return (
        <Box
          className={className}
          id={`${id}-wrapper`}
          ref={ref}
          __css={styles}
          {...rest}
        >
          {headingText && <Heading id={`${id}-heading`} text={headingText} />}
          {descriptionText && <Text>{descriptionText}</Text>}
          {tooltipLabelText ? (
            <Tooltip content={tooltipLabelText}>{children}</Tooltip>
          ) : (
            <>{children}</>
          )}
          <HelperErrorText
            id={`${id}-helperText`}
            isInvalid={isInvalid}
            isRenderedText={showHelperInvalidText}
            text={footnote}
            __css={{ ...styles.helperErrorText, ...helperTextStyles }}
          />
        </Box>
      );
    }
  )
);

export default ComponentWrapper;
