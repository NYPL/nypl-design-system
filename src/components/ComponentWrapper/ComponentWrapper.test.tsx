import * as React from "react";
import { render, screen } from "@testing-library/react";
import { axe } from "jest-axe";
import renderer from "react-test-renderer";

import ComponentWrapper from "./ComponentWrapper";

/**
 * @NOTE while the component is accessible, whatever is passed as children
 * elements may not be considered accessible. This is up to the developer to
 * make sure to pass accessible elements.
 */
describe("ComponentWrapper Accessibility", () => {
  it("Passes axe accessibility test", async () => {
    const { container } = render(
      <ComponentWrapper
        descriptionText="description text"
        headingText="heading text"
        helperText="helper text"
        id="id"
        invalidText="invalid text"
        isInvalid={false}
      >
        <div>children elements</div>
      </ComponentWrapper>
    );
    expect(await axe(container)).toHaveNoViolations();
  });

  it("Passes axe accessibility test without certain props", async () => {
    const { container } = render(
      <ComponentWrapper id="id">
        <div>children elements</div>
      </ComponentWrapper>
    );
    expect(await axe(container)).toHaveNoViolations();
  });
});

describe("ComponentWrapper", () => {
  it("renders with different text sections and children elements", () => {
    render(
      <ComponentWrapper
        descriptionText="description text"
        headingText="heading text"
        helperText="helper text"
        id="id"
        invalidText="invalid text"
        isInvalid={false}
      >
        <div>children elements</div>
      </ComponentWrapper>
    );

    expect(screen.getByText("description text")).toBeInTheDocument();
    expect(screen.getByText("heading text")).toBeInTheDocument();
    expect(screen.getByText("helper text")).toBeInTheDocument();
    expect(screen.queryByText("invalid text")).not.toBeInTheDocument();
    expect(screen.getByText("children elements")).toBeInTheDocument();
  });

  it("renders with only its children elements", () => {
    render(
      <ComponentWrapper id="id">
        <div>children elements</div>
      </ComponentWrapper>
    );

    expect(screen.queryByText("description text")).not.toBeInTheDocument();
    expect(screen.queryByText("heading text")).not.toBeInTheDocument();
    expect(screen.queryByText("helper text")).not.toBeInTheDocument();
    expect(screen.queryByText("invalid text")).not.toBeInTheDocument();
    expect(screen.getByText("children elements")).toBeInTheDocument();
  });

  it("renders invalid text in the internal HelperErrorText component", () => {
    const { rerender } = render(
      <ComponentWrapper
        descriptionText="description text"
        headingText="heading text"
        helperText="helper text"
        id="id"
        invalidText="invalid text"
        isInvalid={false}
      >
        <div>children elements</div>
      </ComponentWrapper>
    );

    expect(screen.getByText("helper text")).toBeInTheDocument();
    expect(screen.queryByText("invalid text")).not.toBeInTheDocument();

    rerender(
      <ComponentWrapper
        descriptionText="description text"
        headingText="heading text"
        helperText="helper text"
        id="id"
        invalidText="invalid text"
        isInvalid={true}
      >
        <div>children elements</div>
      </ComponentWrapper>
    );

    expect(screen.queryByText("helper text")).not.toBeInTheDocument();
    expect(screen.getByText("invalid text")).toBeInTheDocument();
  });

  it("Renders the UI snapshot correctly", () => {
    const basic = renderer
      .create(
        <ComponentWrapper
          descriptionText="description text"
          headingText="heading text"
          helperText="helper text"
          id="id"
          invalidText="invalid text"
          isInvalid={false}
        >
          <div>children elements</div>
        </ComponentWrapper>
      )
      .toJSON();
    const onlyWithChildren = renderer
      .create(
        <ComponentWrapper id="id">
          <div>children elements</div>
        </ComponentWrapper>
      )
      .toJSON();
    const isInvalid = renderer
      .create(
        <ComponentWrapper
          descriptionText="description text"
          headingText="heading text"
          helperText="helper text"
          id="id"
          invalidText="invalid text"
          isInvalid={true}
        >
          <div>children elements</div>
        </ComponentWrapper>
      )
      .toJSON();

    expect(basic).toMatchSnapshot();
    expect(onlyWithChildren).toMatchSnapshot();
    expect(isInvalid).toMatchSnapshot();
  });
});
