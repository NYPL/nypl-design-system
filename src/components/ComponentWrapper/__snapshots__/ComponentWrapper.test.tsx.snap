// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ComponentWrapper Renders the UI snapshot correctly 1`] = `
<div
  className="css-1xdhyk6"
  id="id-wrapper"
>
  <h2
    className="chakra-heading css-1xdhyk6"
    id="id-heading"
  >
    heading text
  </h2>
  <p
    className="chakra-text css-1xdhyk6"
  >
    description text
  </p>
  <div>
    children elements
  </div>
  <div
    aria-atomic={true}
    aria-live="polite"
    className="css-1xdhyk6"
    data-isinvalid={false}
    id="id-helperText"
  >
    <span
      dangerouslySetInnerHTML={
        Object {
          "__html": "helper text",
        }
      }
    />
  </div>
</div>
`;

exports[`ComponentWrapper Renders the UI snapshot correctly 2`] = `
<div
  className="css-1xdhyk6"
  id="id-wrapper"
>
  <div>
    children elements
  </div>
  <div
    aria-atomic={true}
    aria-live="polite"
    className="css-1xdhyk6"
    data-isinvalid={false}
    id="id-helperText"
  />
</div>
`;

exports[`ComponentWrapper Renders the UI snapshot correctly 3`] = `
<div
  className="css-1xdhyk6"
  id="id-wrapper"
>
  <h2
    className="chakra-heading css-1xdhyk6"
    id="id-heading"
  >
    heading text
  </h2>
  <p
    className="chakra-text css-1xdhyk6"
  >
    description text
  </p>
  <div>
    children elements
  </div>
  <div
    aria-atomic={true}
    aria-live="polite"
    className="css-1xdhyk6"
    data-isinvalid={true}
    id="id-helperText"
  >
    <span
      dangerouslySetInnerHTML={
        Object {
          "__html": "invalid text",
        }
      }
    />
  </div>
</div>
`;

exports[`ComponentWrapper Renders the UI snapshot correctly 4`] = `
<div
  className="css-kle7zy"
  id="chakra-wrapper"
>
  <div>
    children elements
  </div>
  <div
    aria-atomic={true}
    aria-live="polite"
    className="css-1xdhyk6"
    data-isinvalid={false}
    id="chakra-helperText"
  />
</div>
`;

exports[`ComponentWrapper Renders the UI snapshot correctly 5`] = `
<div
  className="css-1xdhyk6"
  data-testid="props"
  id="props-wrapper"
>
  <div>
    children elements
  </div>
  <div
    aria-atomic={true}
    aria-live="polite"
    className="css-1xdhyk6"
    data-isinvalid={false}
    id="props-helperText"
  />
</div>
`;
